String implementation.

mholovaty opened the pull request:
Complete String implementation.

mholovaty committed:
Implement basic initialization.

mholovaty committed:
Add todo notes.

mholovaty committed:
Fix iterators constructor to use trailing zero.

Commit:
Fix square brakets in char array allocation.

mholovaty committed:
Make cheap default constructor. Use a shared pointer.

mholovaty committed:
Merge branch 'master' of https://github.com/mholovaty/c-excercises

Conflicts:
	string/string.cc

mholovaty committed:
Fix merge.

mholovaty committed:
Implement constant iterators, equality, deap-copy constructor. Use std::copy_n

mholovaty committed:
Add typdef on std::string::iterator. Enable all tests

mholovaty committed:
Remove direct calls to a smart pointer.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,113 @@
+const char* EMPTY = "";
+
+
+size_t strlen(const char* s)
```
The library provides you with well-tested and super-optimized std::strlen. Do not reinvent the wheel.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,113 @@
+}
+
+
+String::String() {}
```
Or, you can just say:

```
String::String() = default;
```

This seems more explicit to me.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,113 @@
+{
+	mSize = o.size();
+
+	mValue = char_ptr(new char[mSize + 1]);
```
You can use unique_ptr::reset() here.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,113 @@
+
+	std::copy_n(o.begin(), mSize, begin());
+
+	mValue[mSize + 1] = '\0';
```
You may just copy mSize+1 chars. Yoyu are copying from your own String, so you know that it contains the null-terminator.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,113 @@
+
+	mValue = char_ptr(new char[mSize + 1]);
+
+	std::copy_n(first, mSize, begin());
```
Here you can't assume that the terminating zero is here.
Please note: this constructor, ans the other one share almost all of the code. Maybe you could implement the other ones in terms of this?
(Google: delegating constructor)

maciekgajewski replied:
Moving constructor and assignment operators are missing.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,113 @@
+
+bool String::operator==(const char* s) const
+{
+	if (empty() and s == '\0') return true;
```
Oops! You are comparing _s_ (of type _const char_) to _'\0'_ (of type _char_).

This may work, but this probably not what you wanted to implement. I'm pretty sure your compiler gives a warning here.
You can assume that 's' is not null.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,113 @@
+
+std::ostream& operator<<(std::ostream& stream, const String& s)
+{
+	for(auto c: s) stream << c;
```
Slow, but correct.

maciekgajewski commented on string/tests.cc:
```diff
@@ -32,68 +32,66 @@ String IntToString(int i)
 	assert(ss.str() == "12345");
 }
 
-// TODO add more tests here. Achieve 100% test coverage
```
TODO: add more tests. Are all methods covered by unit tests?

mholovaty committed:
Add empty string test

mholovaty committed:
Use 'default' keyword for constructors.
Use overloaded constructors.
Fix char pointer equality comparison.
Use reset for smart pointer initialization.

mholovaty committed:
Complete implementation of String.

mholovaty committed:
Fix style.

mholovaty commented on string/string.cc:
```diff
@@ -1,12 +1,154 @@
+}
+
+// I don't see any other effective way to concatenate for the given API.
+String::String(const String& a, const String& b)
```
Is there a smart concatenation without copying each time?
In Java all  '+' are converted to a StringBuffer by compiler. What's the point of '+' if you need to use a buffer anyway?

maciekgajewski replied:
And string buffer doesn't contain a, uhm, a buffer to which the strings have to be copied?

For 2 strings there is no reasonable optimization. You may want to try to optimize an expression like this:

```
String s = a + b + c + d;
```

by allocating buffer only once. Look for 'expression templates'.

Also, you could optimize operations like this:

```
s += x;
```

by having a pre-allocated buffer in s bigger then needed. This would incur some initial cost, and would allow for avoiding allocation in some cases.

maciekgajewski commented on string/string.h:
```diff
@@ -117,6 +119,14 @@ class String
 	// Compares with c-string. Assumes s is not null
 	bool operator==(const char* s) const;
+
+protected:
```
use 'private'. 'protected' is rarely needed, and even more rarely correct.

maciekgajewski commented on string/string.cc:
```diff
@@ -1,12 +1,154 @@
+String::String(const char* s): String(s, s + std::strlen(s)) {}
+
+template String::String(std::string::iterator first, std::string::iterator last);
+
```
The definition of function template should be in the header file, so that all clients can 'see' it, and it would be automatically instantiated for all types required.
The the line above, instantiating for std::string::iterator wouldn't be needed.

mholovaty committed:
Move String template constructor to a header.
Change implementation members to be private instead of protected

mholovaty committed:
Implement a simple arguments sorting command.

mholovaty committed:
Add case-insensitive comparison.

mholovaty committed:
Add empty plot sub-project.

mholovaty committed:
Add plotting interface.

mholovaty committed:
Add plotting interface.

mholovaty committed:
Use any_range type for plotting interface.

mholovaty committed:
Add what I have at the moment.

mholovaty committed:
Add plot range function.

mholovaty committed:
Change extention of engine files to cc.

mholovaty committed:
Delete engine_gnuplot.h

mholovaty committed:
Delete engine_qt.h

mholovaty committed:
Delete engine_csv.h

mholovaty committed:
moo/CMakeLists.txt

mholovaty committed:
Add cmake file for moo

mholovaty committed:
Merge branch 'master' of https://github.com/mholovaty/c-excercises

mholovaty committed:
Remove unused include

mholovaty committed:
Update expression parser to use the latest Spirit API.

mholovaty committed:
Add a hacky implemenation of expressions parser.

mholovaty committed:
Fix namespace.

mholovaty committed:
Add a simple parser implementation.

mholovaty committed:
Implement optional expression and term multiplicators.

mholovaty committed:
Remove debug prints.

mholovaty committed:
Fix formatting.

mholovaty committed:
Implement AST eval.

mholovaty committed:
Update AST to use unique_ptr for node links.

mholovaty committed:
Add EvalError

mholovaty committed:
Remove undefined node type

mholovaty committed:
Fix formatting.

mholovaty committed:
Move spirit related implementation into the same namespace.
